contracts/StoryProtocolGateway.sol

```diff
@@ -1,503 +1,424 @@
+        ipId = IP_ASSET_REGISTRY.register(block.chainid, nftContract, tokenId);
+        _setMetadata(metadata, ipId);
+
+        licenseTermsId = _registerPILTermsAndAttach(ipId, terms);
+        ISPGNFT(nftContract).safeTransferFrom(address(this), recipient, tokenId, "");
     }
 
-    /// @notice Add a PIL policy to an IPAsset. Create a new PIL policy if it doesn't exist.
-    /// @param pilPolicy The PIL policy to add to the Licensing Module.
-    /// @param transferable Whether or not the license is transferable.
-    /// @param royaltyPolicy Address of a royalty policy contract (e.g. RoyaltyPolicyLAP) that will handle royalty
-    /// payments.
-    /// @param mintingFee Fee to be paid when minting a license.
-    /// @param mintingFeeToken Token to be used to pay the minting fee.
-    /// @param ipId The address of the IP asset to add the policy to.
-    /// @return policyGlobalId The ID of the newly (or existing) registered policy.
-    /// @return policyIndexOnIpId The index of the policy on the IP asset.
-    function addPILPolicyToIp(
-        PILPolicy memory pilPolicy,
-        bool transferable,
-        address royaltyPolicy,
-        uint256 mintingFee,
-        address mintingFeeToken,
-        address ipId
-    ) external returns (uint256 policyGlobalId, uint256 policyIndexOnIpId) {
-        policyGlobalId = PIL_POLICY_FRAMEWORK_MANAGER.registerPolicy(
-            RegisterPILPolicyParams({
-                transferable: transferable,
-                royaltyPolicy: royaltyPolicy,
-                mintingFee: mintingFee,
-                mintingFeeToken: mintingFeeToken,
-                policy: pilPolicy
-            })
+    /// @notice Register a given NFT as an IP and attach Programmable IP License Terms.
+    /// @dev Because IP Account is created in this function, we need to set the permission via signature to allow this
+    /// contract to attach PIL Terms to the newly created IP Account in the same function.
+    /// @param nftContract The address of the NFT collection.
+    /// @param tokenId The ID of the NFT.
+    /// @param metadata OPTIONAL. The desired metadata for the newly registered IP.
+    /// @param terms The PIL terms to be registered.
+    /// @param sigMetadata OPTIONAL. Signature data for setAll (metadata) for the IP via the Core Metadata Module.
+    /// @param sigAttach Signature data for attachLicenseTerms to the IP via the Licensing Module. The nonce of this
+    /// signature must be one above `sigMetadata` if the metadata is being set, ie. `sigMetadata` is non-empty.
+    /// @return ipId The ID of the registered IP.
+    /// @return licenseTermsId The ID of the registered PIL terms.
+    function registerIpAndAttachPILTerms(
+        address nftContract,
+        uint256 tokenId,
+        IPMetadata calldata metadata,
+        PILTerms calldata terms,
+        SignatureData calldata sigMetadata,
+        SignatureData calldata sigAttach
```

Comment:
The two SignatureData means SDK user have to sign twice.
It would be better to merge the two signatures into one by leveraging the `setBatchPermissions()` of AccessController.

Reply:
I agree, I've been working on that in the next PR as a usability improvement. This PR is pretty big for more changes.